### PR TITLE
[IOPLT-236] Adds the new `OTPInput` component

### PR DESCRIPTION
**Pull request by CrisTofani**
## Short description
This PR implements the new `OTPInput` component.

## List of changes proposed in this pull request
- `BoxedInput` component for number/secret representation
- `OTPInput` component to mask a real hidden TextInput

## How to test
Check the example app related page

https://github.com/pagopa/io-app-design-system/assets/3959405/75b3fc97-c242-4080-ae27-ffde37008aca

### CODE REVIEW

**Review by StefanoFrontini — COMMENTED** [review]
When clicking on an input box and then pressing a keybord key  nothing happens (only the , and . seems to work)

**Comment by CrisTofani** [comment]
> When clicking on an input box and then pressing a keybord key nothing happens (only the , and . seems to work)

addressed in https://github.com/pagopa/io-app-design-system/pull/161/commits/440757f44c016b8841414693e015ed91e27b5d84

File: src/components/otpInput/OTPInput.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        const timer = setTimeout(() => {
+          setHasError(false);
+          setInputValue("");
+          onValueChange("");
+        }, 500);
+        return () => clearTimeout(timer);
+      }
+    }
+    return;
+  }, [value, onValidate, length, onValueChange, translate, shakeAnimation]);
```
Comment (StefanoFrontini) [inline src/components/otpInput/OTPInput.tsx:79]:
```suggestion
const inputRef = useRef<TextInput>(null);

  const timerRef = useRef<NodeJS.Timeout>();
  const handleValidate = (val: string) => {
    if (!onValidate || val.length < length) {
      return;
    }

    const isValid = onValidate(val);

    if (!isValid) {
      setHasError(true);
      triggerHaptic("notificationError");

      // eslint-disable-next-line functional/immutable-data
      translate.value = shakeAnimation();

      if (timerRef.current) {
        clearTimeout(timerRef.current);
      }

      // eslint-disable-next-line functional/immutable-data
      timerRef.current = setTimeout(() => {
        setHasError(false);
        setInputValue("");
        onValueChange("");
      }, 500);
    }
  };
  const handleChange = (value: string) => {
    if (value.length > length) {
      return;
    }
    setInputValue(value);
    onValueChange(value);
    handleValidate(value);
  };

`

File: src/components/otpInput/OTPInput.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import { useEffect, useState } from "react";
```
Comment (StefanoFrontini) [inline src/components/otpInput/OTPInput.tsx:2]:
```suggestion
import { useRef, useState } from "react";
```

File: src/components/otpInput/OTPInput.tsx (hunk 1 above)
Comment (CrisTofani) [reply src/components/otpInput/OTPInput.tsx:79]: I'm just missing one point on this suggestion, what would trigger the validation callback?

Comment (StefanoFrontini) [reply src/components/otpInput/OTPInput.tsx:79]: It should be triggered by the handleChange function

Comment (CrisTofani) [reply src/components/otpInput/OTPInput.tsx]: Tested out, it actually works perfectly!

Just some changes:
- I left the `createRef` for `inputRef` cause of type specifications
- I left the `useEffect` to handle the timeout clear when the component is unmounted

Comment (CrisTofani) [reply src/components/otpInput/OTPInput.tsx]: [513c46a](https://github.com/pagopa/io-app-design-system/pull/161/commits/513c46afa9a0aff1ba79b7afdffaab70fa2b6946)

**Review by StefanoFrontini — APPROVED** [review]

**Review by dmnplb — APPROVED** [review]

**Pull request merged**